### PR TITLE
Introduce the SimpleStateChecker.

### LINKED ISSUE #8679: Add Pemission & Feature & GlobalFeatures RequiredPermissionName to ApplicationMenuItem & ToolbarItem
Same with https://github.com/abpframework/abp/pull/8543

Related https://github.com/abpframework/abp/pull/7799

### PR DESCRIPTION
**Pull request by maliming**
Resolve #8679

Any class can inherit `IHasSimpleStateCheckers `to support state checks. We have enabled status checks for permission, menu, and toolbar. They will be disabled/enabled based on permissions or features, etc.

You need to add a checker that implements `ISimpleStateChecker` to the instance's `SimpleStateCheckers`.

I have built-in general authenticated/permissions/features/global feature state checks.
Instances that inherit from `IHasSimpleStateCheckers` can call these extension methods.

```cs
RequireAuthenticated();
RequirePermissions(bool requiresAll, params string[] permissions);
RequireFeatures(bool requiresAll, params string[] features);
RequireGlobalFeatures(bool requiresAll, params Type[] globalFeatures);
```

If there are many instance items that require state checking, there may be performance problems.

In this case, you can implement `ISimpleBatchStateChecker`. It can check multiple items at once. 
You need to make sure that the same `ISimpleBatchStateChecker` instance is added to the `SimpleStateCheckers` of multiple instances.

The `IsEnabledAsync` method of `ISimpleStateCheckerManager` supports single and batch checks.

```cs
Task<bool> IsEnabledAsync(TState state);
Task<SimpleStateCheckerResult<TState>> IsEnabledAsync(TState[] states);
```

### CODE REVIEW

**Review by enisn — APPROVED** [review]
LGTM

File: framework/src/Volo.Abp.Authorization.Abstractions/Volo/Abp/Authorization/Permissions/PermissionDefinition.cs [hunk 1]
```diff
@@ -31,7 +32,7 @@ public class PermissionDefinition
         /// </summary>
         public List<string> Providers { get; } //TODO: Rename to AllowedProviders?
 
-        public List<IPermissionStateProvider> StateProviders { get; }
+        public List<ISimpleStateChecker<PermissionDefinition>> SimpleStateCheckers { get; }
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Authorization.Abstractions/Volo/Abp/Authorization/Permissions/PermissionDefinition.cs:35]: Rename property to `StateCheckers` (keep `ISimpleStateChecker` naming).

File: framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/PermissionChecker.cs [hunk 2]
```diff
@@ -15,20 +16,20 @@ public class PermissionChecker : IPermissionChecker, ITransientDependency
         protected ICurrentPrincipalAccessor PrincipalAccessor { get; }
         protected ICurrentTenant CurrentTenant { get; }
         protected IPermissionValueProviderManager PermissionValueProviderManager { get; }
-        protected IPermissionStateManager PermissionStateManager { get; }
+        protected ISimpleStateCheckerManager<PermissionDefinition> PermissionSimpleStateCheckerManager { get; }
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/PermissionChecker.cs:19]: `PermissionSimpleStateCheckerManager` -> `StateCheckerManager` renaming would be simpler.

File: framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/PermissionSimpleStateCheckerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            where TState : IHasSimpleStateCheckers<TState>
+        {
+            state.RequirePermissions(requiresAll: requiresAll, batchCheck: true, permissions);
+            return state;
+        }
+
+        public static TState RequirePermissions<TState>(
+            [NotNull] this TState state,
+            bool requiresAll,
+            bool batchCheck = true,
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/PermissionSimpleStateCheckerExtensions.cs:37]: Making this with default value is confusing.

File: framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/PermissionSimpleStateCheckerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            params string[] permissions)
+            where TState : IHasSimpleStateCheckers<TState>
+        {
+            Check.NotNull(state, nameof(state));
+            Check.NotNullOrEmpty(permissions, nameof(permissions));
+
+            if (batchCheck)
+            {
+                lock (state)
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/PermissionSimpleStateCheckerExtensions.cs:46]: Why we are locking the state? I didn't understand that. This code already should not be called in multiple threads.

File: framework/src/Volo.Abp.Core/Volo/Abp/SimpleStateChecking/AbpSimpleStateCheckerOptions.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿using Volo.Abp.Collections;
+
+namespace Volo.Abp.SimpleStateChecking
+{
+    public class AbpSimpleStateCheckerOptions<TState>
+        where TState : IHasSimpleStateCheckers<TState>
+    {
+        public ITypeList<ISimpleStateChecker<TState>> GlobalSimpleStateCheckers { get; }
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Core/Volo/Abp/SimpleStateChecking/AbpSimpleStateCheckerOptions.cs:8]:
```suggestion
        public ITypeList<ISimpleStateChecker<TState>> GlobalStateCheckers { get; }
```

**Pull request merged**